README.md

L2+
React Native app for testing WebView on diffrent APIs.

L26 /a.
Need more tests for HTML alert 'cause
doesn't work on next devices/emulators:
  device api 22 (Cube)
  emulator api 21
  emulator api 22

works on next devices/emulators:
  device api 21 (Samsung)
  device api 23 (Meizu)
  device api 25 (Xiaomi 4X)
  emulator api 23
  emulator api 28
  emulator api 29
  emulator api 30

L26 /b.
Test of HTML body_tag.scrollTo({top: position})
doesn't work on:
  device api 22 (Cube)
  emulator api 21
  emulator api 22
  emulator api 23

works on:
  device api 25 (Xiaomi 4X)
  device api 21 (Samsung)
  emulator api 28
  emulator api 29
  emulator api 30

L26 /c.
Test of HTML element.scrollIntoView(...)
Seems it works everywhere. Tested on:
  device api 22 (Cube)
  emulator api 21
  emulator api 22
  emulator api 23
  emulator api 28
behavior: "smooth" isn't on:
  device api 23 (Meizu)
  device api 22 (Cube)
  emulators <= api 23
but works on:
  device api 21 (Samsung)

                      | alert | scr1 | scr2 | smooth
Devices:              |       |      |      |
  api 21 (Samsung)    |   +   |  +   |  +   |  +
  api 22 (Cube)       |   -   |  -   |  +   |  -
  api 23 (Maizu)      |   -   |  +   |  +   |  -
  api 25 (Xiaomi 4X)  |   +   |  +   |  +   |  +
  api 28 (Android TV) |   ?   |  ?   |  ?   |  ?
Emulators:            |       |      |      |
  api 21              |   -   |  -   |  +   |  -
  api 22              |   -   |  -   |  +   |  -
  api 23              |   -   |  +   |  +   |  -
  api 28              |   +   |  +   |  +   |  +
  api 29              |   +   |  +   |  +   |  +
  api 30              |   +   |  +   |  +   |  +

Result apks:
5 190 502 app-arm64-v8a-release.apk
4 846 043 app-armeabi-v7a-release.apk
5 470 715 app-x86_64-release.apk
5 574 979 app-x86-release.apk


ss/releases
#/usr/bash

cd ../android

echo ' => build release apks...'
./gradlew assembleRelease
echo 'gradlew DONE.'  


libs/mxlib.ts
//==================================================================================================
//  Library for Maix common functions (KAJ)
//==================================================================================================

export const _none_ = (_anything?: any) => {};
export const cccc = _none_;
export const _ccc = _none_;
export const _cc = _none_;
export const cc = console.log;
export const ccc = (t: string) => cc(getTimeNow() + ': ' + t);
export const cct = (...args: any[]) => cc('[' + getTimeNow() + ']', ...args);
export const secx = () => getTimeNow().substr(-6);
export const pref0 = (n: number, len: number): string => ('0'.repeat(len) + n).slice(-1 * len);
export const pref0_ = (n: number, len: number): string =>
  String(n).length < n ? ('0'.repeat(len) + n).slice(-1 * len) : String(n);
export const EOL = '\n';
export const LL = '\n';
export const TAB = '\t';
export const Q = "'";
export const QQ = '"';
export const BR = '<br/>';

export const unixMilisec = () => new Date().getTime();
export const sFTime = (time?: string): string => {
  const d = time ? new Date(time) : new Date();
  const _day = _add2(d.getDate());
  const _month = _add2(d.getMonth() + 1);
  const _year = d.getFullYear();
  return `${_year}.${_month}.${_day} ` + getTimeMx(d);
  //return oDateNow.toLocaleString() + '.' + _add3(oDateNow.getMilliseconds());
};
export const getTimeMx = (objDate: Date): string => {
  let res = '';
  res += _add2(objDate.getHours()) + ':';
  res += _add2(objDate.getMinutes()) + ':';
  res += _add2(objDate.getSeconds());
  res += '.' + _add3(objDate.getMilliseconds());
  return res;
};
const _add3 = (n: number): string => pref0(n, 3);
const _add2 = (n: number): string => pref0(n, 2);

export const getTimeNow = (): string => getTimeMx(new Date());

export const T = () => getTimeMx(new Date());
export const dT = (st: Date, end = new Date()): number => end.getTime() - st.getTime();
export const dTs = (st: Date, end = new Date()): string =>
  '{Δ:' + (end.getTime() - st.getTime()) + 'ms}';

export const YorN = (b: boolean) => (b ? 'YES' : 'no');
export const yn = (ok: boolean): string => (ok ? '+' : '-');

//--------------------------------------------------------------------------------------------------
export const getRandomInt = (max: number): number => Math.floor(Math.random() * Math.floor(max));

export const f = async (msg: string, ms: number) => {
  let pattern = `delay "${msg}" with ${ms}`;
  cc(sFTime() + `: ${pattern} starts...`);
  let promise = new Promise((resolve) => {
    setTimeout(() => resolve(msg), ms);
  });
  let result = await promise;
  cc(sFTime() + `: ${pattern} ends with result: "${result}".`);
};

export const J = (obj: any): string => JSON.stringify(obj, null, '  ');
export const lObj = (obj: any) => cc(JSON.stringify(obj, null, '    '));
export const JObj = (obj: any, tabs?: string) => cc(JSON.stringify(obj, null, tabs || '  '));
export const lObj_ = (obj: any) => cc(JSON.stringify(obj, null, '    '));

export const numbBy3 = (numb: number): string => numb.toString().replace(/\d(?=(\d{3})+$)/g, '$& ');
export const numbBy3s = (numb: string): string => {
  // Error with RNFS.ReadDirItem.size !!! Need deeper checks.
  if (numb.length < 4) {
    return numb;
  }
  return numb.trim().replace(/\d(?=(\d{3})+$)/g, '$& ');
};

export const b2Kb = (bytes: number): number => bytes / 1024.0;
export const b2Mb = (bytes: number): number => bytes / (1024 * 1024.0);
export const b2Gb = (bytes: number): number => bytes / (1024 * 1024 * 1024.0);

export const lSort = (a: string, b: string): number =>
  a.localeCompare(b, undefined, {sensitivity: 'base'});

export const uniqueArray = (arr: any[]): any[] => [...new Set(arr)];

export const findIndexAny = (val: any, arr: any[]): number =>
  arr.findIndex((el: any) => el === val);
export function findIndexF<_T>(val: _T, arr: _T[]): number {
  return arr.findIndex((el: _T) => el === val);
}
export const findIndexT = <_T>(val: _T, arr: _T[]): number => arr.findIndex((el: _T) => el === val);

export const lHash = (hash: any) => hash.forEach(_logMapElements);
const _logMapElements = (value: any, key: any) => cc(`{${key}} => "${value}"`);

export const aSortUnique = (arr: string[]): string[] => [...new Set(arr)].sort();
export const aSortUniqueNumb = (arr: number[]): number[] => [...new Set(arr)].sort((a, b) => a - b);

export const aArrayUnion = (arr1: string[], arr2: string[]): string[] => arr1.concat(arr2);

export const sGetByREx = (str: string, re: string): string => {
  const regex = new RegExp(re, 'g');
  let match: any;
  match = regex.exec(str);
  return match && match.length >= 2 ? match[1] : '';
};

export const aGetByREx = (str: string, re: string): string[] => {
  let aRes: string[] = [];
  const regex = new RegExp(re, 'g');
  const match = regex.exec(str) || [];
  if (match.length >= 2) {
    aRes.push(match[1]);
  }
  return aRes;
};

export const showProps = (obj: any, objName: any) => {
  var result = '';
  for (var i in obj) {
    // obj.hasOwnProperty() is used to filter out properties from the object's prototype chain
    if (obj.hasOwnProperty(i)) {
      result += `${objName}.${i} = ${obj[i]}\n`;
    }
  }
  return result;
};

export const replaceAll = (str: string, find: string, replace: string) => {
  return str.replace(new RegExp(find, 'g'), replace);
};

//--------------------------------------------------------------------------------------------------
// eslint-disable-next-line no-shadow
enum OBJTYPE {
  FUNC = 'function',
  OBJ = 'object',
  ANY = 'any',
}
interface IObj {
  type: OBJTYPE;
  isObj: boolean;
  value: string;
}
// show more extended object's properties, where JSON.stringify can miss, with some counters
interface IShowObjResult {
  message: string;
  counter: number;
}
export const showOBJ = (Obj: any, index = ''): IShowObjResult => {
  const keysArray: string[] = [];
  const propsMap = new Map<string, IObj>();
  for (const property in Obj) {
    keysArray.push(property);
    const nowObj = Obj[property];
    const strType = typeof nowObj;
    //cc('nowObj, strType', nowObj, strType)
    const type: OBJTYPE =
      strType === OBJTYPE.OBJ ? OBJTYPE.OBJ : strType === OBJTYPE.FUNC ? OBJTYPE.FUNC : OBJTYPE.ANY;
    const value = type === OBJTYPE.ANY ? Obj[property] : strType;
    propsMap.set(property, {
      type: type,
      isObj: type === OBJTYPE.OBJ,
      value: value,
    });
  }
  keysArray.sort();
  let i = 0;
  let counter = 0;
  keysArray.forEach((key: string) => {
    const _props = propsMap.get(key);
    if (_props) {
      i++;
      counter++;
      let _value = _props.value;
      if (_props.type === OBJTYPE.OBJ) {
        _value = '[' + _value.toUpperCase() + ']';
      }
      if (_props.type === OBJTYPE.FUNC) {
        _value += '()';
      }
      if (_props.type === OBJTYPE.ANY) {
        _value = QQ + _value + QQ;
      }
      cc(`${index}${i}. "${key}": ${_value}`);
      if (_props.isObj) {
        const _res: IShowObjResult = showOBJ(Obj[key], i + '.');
        counter += _res.counter;
      }
    }
  });
  const res: IShowObjResult = {message: 'showOBJ. Total properties = ' + counter, counter: counter};
  return res;
};


android/app/src/main/java/com/webviewl1u/MainApplication.java
package com.webviewl1u;

import android.app.Application;
import android.content.Context;
import com.facebook.react.PackageList;
import com.facebook.react.ReactApplication;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.ReactPackage;
import com.facebook.soloader.SoLoader;
import java.lang.reflect.InvocationTargetException;
import java.util.List;

public class MainApplication extends Application implements ReactApplication {

  private final ReactNativeHost mReactNativeHost =
      new ReactNativeHost(this) {
        @Override
        public boolean getUseDeveloperSupport() {
          return BuildConfig.DEBUG;
        }

        @Override
        protected List<ReactPackage> getPackages() {
          @SuppressWarnings("UnnecessaryLocalVariable")
          List<ReactPackage> packages = new PackageList(this).getPackages();
          // Packages that cannot be autolinked yet can be added manually here, for example:
          // packages.add(new MyReactNativePackage());
          return packages;
        }

        @Override
        protected String getJSMainModuleName() {
          return "index";
        }
      };

  @Override
  public ReactNativeHost getReactNativeHost() {
    return mReactNativeHost;
  }

  @Override
  public void onCreate() {
    super.onCreate();
    SoLoader.init(this, /* native exopackage */ false);
    initializeFlipper(this, getReactNativeHost().getReactInstanceManager());
  }

  /**
   * Loads Flipper in React Native templates. Call this in the onCreate method with something like
   * initializeFlipper(this, getReactNativeHost().getReactInstanceManager());
   *
   * @param context
   * @param reactInstanceManager
   */
  private static void initializeFlipper(
      Context context, ReactInstanceManager reactInstanceManager) {
    if (BuildConfig.DEBUG) {
      try {
        /*
         We use reflection here to pick up the class that initializes Flipper,
        since Flipper library is not available in release mode
        */
        Class<?> aClass = Class.forName("com.webviewl1u.ReactNativeFlipper");
        aClass
            .getMethod("initializeFlipper", Context.class, ReactInstanceManager.class)
            .invoke(null, context, reactInstanceManager);
      } catch (ClassNotFoundException e) {
        e.printStackTrace();
      } catch (NoSuchMethodException e) {
        e.printStackTrace();
      } catch (IllegalAccessException e) {
        e.printStackTrace();
      } catch (InvocationTargetException e) {
        e.printStackTrace();
      }
    }
  }
}


android/app/src/debug/java/com/webviewl1u/ReactNativeFlipper.java
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * <p>This source code is licensed under the MIT license found in the LICENSE file in the root
 * directory of this source tree.
 */
package com.webviewl1u;

import android.content.Context;
import com.facebook.flipper.android.AndroidFlipperClient;
import com.facebook.flipper.android.utils.FlipperUtils;
import com.facebook.flipper.core.FlipperClient;
import com.facebook.flipper.plugins.crashreporter.CrashReporterPlugin;
import com.facebook.flipper.plugins.databases.DatabasesFlipperPlugin;
import com.facebook.flipper.plugins.fresco.FrescoFlipperPlugin;
import com.facebook.flipper.plugins.inspector.DescriptorMapping;
import com.facebook.flipper.plugins.inspector.InspectorFlipperPlugin;
import com.facebook.flipper.plugins.network.FlipperOkhttpInterceptor;
import com.facebook.flipper.plugins.network.NetworkFlipperPlugin;
import com.facebook.flipper.plugins.react.ReactFlipperPlugin;
import com.facebook.flipper.plugins.sharedpreferences.SharedPreferencesFlipperPlugin;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.modules.network.NetworkingModule;
import okhttp3.OkHttpClient;

public class ReactNativeFlipper {
  public static void initializeFlipper(Context context, ReactInstanceManager reactInstanceManager) {
    if (FlipperUtils.shouldEnableFlipper(context)) {
      final FlipperClient client = AndroidFlipperClient.getInstance(context);

      client.addPlugin(new InspectorFlipperPlugin(context, DescriptorMapping.withDefaults()));
      client.addPlugin(new ReactFlipperPlugin());
      client.addPlugin(new DatabasesFlipperPlugin(context));
      client.addPlugin(new SharedPreferencesFlipperPlugin(context));
      client.addPlugin(CrashReporterPlugin.getInstance());

      NetworkFlipperPlugin networkFlipperPlugin = new NetworkFlipperPlugin();
      NetworkingModule.setCustomClientBuilder(
          new NetworkingModule.CustomClientBuilder() {
            @Override
            public void apply(OkHttpClient.Builder builder) {
              builder.addNetworkInterceptor(new FlipperOkhttpInterceptor(networkFlipperPlugin));
            }
          });
      client.addPlugin(networkFlipperPlugin);
      client.start();

      // Fresco Plugin needs to ensure that ImagePipelineFactory is initialized
      // Hence we run if after all native modules have been initialized
      ReactContext reactContext = reactInstanceManager.getCurrentReactContext();
      if (reactContext == null) {
        reactInstanceManager.addReactInstanceEventListener(
            new ReactInstanceManager.ReactInstanceEventListener() {
              @Override
              public void onReactContextInitialized(ReactContext reactContext) {
                reactInstanceManager.removeReactInstanceEventListener(this);
                reactContext.runOnNativeModulesQueueThread(
                    new Runnable() {
                      @Override
                      public void run() {
                        client.addPlugin(new FrescoFlipperPlugin());
                      }
                    });
              }
            });
      } else {
        client.addPlugin(new FrescoFlipperPlugin());
      }
    }
  }
}


libs/defOS.ts
/* eslint-disable curly */
import {Platform} from 'react-native';

export const strOS: string = Platform.OS;
export const isANDROID: boolean = Platform.OS === 'android';
export const isWEB: boolean = Platform.OS === 'web';
export const isIOS: boolean = Platform.OS === 'ios';
//export const isELEC: boolean = isWEB && navigator.userAgent.toLowerCase().search('electron') >= 0;
//export const isLINUX: boolean = false;
//export const isWINDOWS: boolean = isWEB && navigator.userAgent.toLowerCase().search('windows') >= 0;

export const YorN = (b: boolean) => (b ? 'YES' : 'no');
export const YN = (ok: boolean): string => (ok ? '+' : '-');

export const nameOS = (): string => {
  let _nameOS = '_UNDEFINED_';
  if (isANDROID) _nameOS = 'ANDROID';
  else if (isIOS) _nameOS = 'IOS';
  //else if (isELEC) _nameOS = isWINDOWS ? 'WINDOWS' : 'LINUX';
  //else if (isWEB) _nameOS = 'WEB';
  return _nameOS;
};

// /* eslint-disable curly */
//==================================================================================================
//  Library for React Native
//==================================================================================================
// import {Platform} from 'react-native';

export const getOSver = (): string => {
  //const verInfo = 'OS:' + Platform.OS;
  //const verInfo = Platform.OS === 'android' ? ' A:' + andAPIlvl() : 'OS:' + Platform.OS;
  return Platform.OS === 'android' ? ' A:' + andAPIlvl() : 'OS:' + Platform.OS;
};

export const andAPIlvl = (): string =>
  Platform.OS === 'android' ? Platform.Version.toString() : '';

export const andAPI = (): number => {
  let res: number = 0;
  if (Platform.OS === 'android') {
    const ver = Platform.Version.toString();
    res = parseInt(ver, 10);
  }
  return res;
};

export const isAnd10plus = () => andAPI() > 28;

const verInfo: Map<number | string, string> = new Map([
  [16, '4.1'],
  [17, '4.2'],
  [18, '4.3'],
  [19, '4.4'],
  [20, '4.4W'],
  [21, '5.0'],
  [22, '5.1'],
  [23, '6.0'],
  [24, '7.0'],
  [25, '7.1'],
  [26, '8.0'],
  [27, '8.1'],
  [28, '9.0'],
  [29, '10.0'],
  [30, '11.0'],
]);

export const andVer = (): string => {
  let res = '';
  if (Platform.OS === 'android') {
    if (Platform.Version > 30) res = '>11.0';
    else if (Platform.Version >= 16) res = verInfo.get(Platform.Version) || '<4.1';
    else res = '<4.1';
  }
  return res;
};

/*
    Data from Android Studio on K4E:
    100.0%      4.0      15
     99.8%      4.1      16
     99.2%      4.2      17
     98.4%      4.3      18
     98.1%      4.4      19
     94.1%      5.0      21
     92.3%      5.1      22
     84.9%      6.0      23
     73.7%      7.0      24
     66.2%      7.1      25
     60.8%      8.0      26
     53.5%      8.1      27
     39.5%      9.0      28
      8.2%     10.0      29

    Data from Android Studio on K3Q:
    100.0%      4.0      15
     99.6%      4.1      16
     98.1%      4.2      17
     95.9%      4.3      18
     95.3%      4.4      19
     85.0%      5.0      21
     80.2%      5.1      22
     62.6%      6.0      23
     37.1%      7.0      24
     14.2%      7.1      25
      6.0%      8.0      26
      1.1%      8.1      27


    Data on K3P
    Android   10.0       29    Q
    Android     9        28    P
    Android     8.1      27    O_MR1
    Android     8.0      26    O
    Android     7.1.1    25    N_MR1
    Android     7.1      25    N_MR1
    Android     7.0      24    N
    Android     6.0      23    M
    Android     5.1      22    LOLLIPOP_MR1
    Android     5.0      21    LOLLIPOP
    Android     4.4W     20    KITKAT_WATCH   KitKat for Wearables Only
    Android     4.4      19    KITKAT
    Android     4.3      18    JELLY_BEAN_MR2
    Android     4.2.2    17    JELLY_BEAN_MR1
    Android     4.2      17    JELLY_BEAN_MR1
    Android     4.1.1    16    JELLY_BEAN
    Android     4.1      16    JELLY_BEAN
    Android     4.0.4    15    ICE_CREAM_SANDWICH_MR1
    Android     4.0.3    15    ICE_CREAM_SANDWICH_MR1
    Android     4.0.2    14    ICE_CREAM_SANDWICH
    Android     4.0.1    14    ICE_CREAM_SANDWICH
    Android     4.0      14    ICE_CREAM_SANDWICH
    Android     3.2      13    HONEYCOMB_MR2
    Android     3.1.x    12    HONEYCOMB_MR1
    Android     3.0.x    11    HONEYCOMB
    Android     2.3.4
    Android     2.3.3    10    GINGERBREAD_MR1
    Android     2.3.2
    Android     2.3.1
    Android     2.3       9    GINGERBREAD
    Android     2.2.x     8    FROYO
    Android     2.1.x     7    ECLAIR_MR1
    Android     2.0.1     6    ECLAIR_0_1
    Android     2.0       5    ECLAIR
    Android     1.6       4    DONUT
    Android     1.5       3    CUPCAKE
    Android     1.1       2    BASE_1_1
    Android     1.0       1    BASE

*/

/*
export const andVer1 = (): string => {
  let res = '';
  if (Platform.OS !== 'android') return '';
  if (Platform.Version > 29)
    res = '>10.0'
  else {
    switch (Platform.Version) {
      case 29:
        res = '10.0';
        break;
      case 28:
        res = '9.0';
        break;
      case 27:
        res = '8.1';
        break;
      case 26:
        res = '8.0';
        break;
      case 25:
        res = '7.1';
        break;
      case 24:
        res = '7.0';
        break;
      case 23:
        res = '6.0';
        break;
      case 22:
        res = '5.1';
        break;
      case 21:
        res = '5.0';
        break;
      case 20:
        res = '4.4W';
        break;
      case 19:
        res = '4.4';
        break;
      case 18:
        res = '4.3';
        break;
      case 17:
        res = '4.2';
        break;
      case 16:
        res = '4.1';
        break;
      default:
        res = '<4.1'
    }
  }
  return res;
}
*/
